<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/assignment/ass6/hpdp/HAWI/HAWISAN_HPDP_ASS6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ASSIGNMENT 6 (Mastering In Handling Big Data)


This dataset provides comprehensive information on property sales in England and Wales, as sourced from the UK government's HM Land Registry. It offers valuable insights into property transactions, including sale prices, locations, and types of properties sold. This dataset is particularly useful for analysts, researchers, and businesses looking to understand market trends, property valuations, and investment opportunities in the real estate sector of England and Wales.



# LOADING THE DATASET AND CLEANING

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import random

In [ ]:
def drop_duplicates(df):
    old = df.shape[0]
    df.drop_duplicates(inplace=True)
    new = df.shape[0]
    count = old - new
    if (count == 0):
        print("No duplicate rows were found.")
    else:
        print(f"{count} duplicate rows were found and removed.")

In [ ]:
from  google.colab import files
files.upload()

Saving kaggle.json to kaggle.json


{'kaggle.json': b'{"username":"izzatkiwi","key":"2a4c21c8c7ca8aa5d22189f638b67c40"}'}

In [ ]:
!pip install kaggle # 1. Install the Kaggle Package
!mkdir -p ~/.kaggle # 2. Create a Kaggle Directory
!cp kaggle.json ~/.kaggle/ # 3. Upload the Kaggle API Token
!chmod 600 ~/.kaggle/kaggle.json # 4. Set File Permission

In [ ]:
!kaggle datasets download lorentzyeung/price-paid-data-202304

 99% 1.34G/1.36G [00:16<00:00, 114MB/s] 
100% 1.36G/1.36G [00:16<00:00, 87.2MB/s]


In [ ]:
! unzip price-paid-data-202304.zip

Archive:  price-paid-data-202304.zip
  inflating: 202304.csv              


In [ ]:
filename = r"202304.csv"
n = sum(1 for line in open(filename)) - 1
chunk_size = 1000000

s = n // 10
skip = sorted(random.sample(range(1, n + 1), n - s))

chunks = []
for chunk in pd.read_csv(filename, chunksize=chunk_size, skiprows=skip):
    chunks.append(chunk)

df = pd.concat(chunks, ignore_index=True)

In [ ]:

columns_list = ['Transaction_unique_identifier', 'price', 'Date_of_Transfer', 'postcode', 'Property_Type', 'Old/New', 'Duration', 'PAON', 'SAON', 'Street', 'Locality', 'Town/City', 'District', 'County', 'PPDCategory_Type', 'Record_Status - monthly_file_only']


df.columns = columns_list

In [ ]:
df

Transaction_unique_identifier   price  Date_of_Transfer  \
0        {003DB740-4F22-46E5-AD3C-52EB15407DE5}   62750  1995-10-02 00:00   
1        {38008828-F33E-4F54-927C-4BB90B83248E}   51000  1995-10-31 00:00   
2        {2A9D6B70-FC5D-46A1-A8EF-5A06AF633839}   51000  1995-02-17 00:00   
3        {4AB9B105-8E30-446E-A3DD-4F4992828E8F}   89500  1995-07-14 00:00   
4        {E58774C2-3011-484A-B631-5A06F3FB5BC9}  102500  1995-09-28 00:00   
...                                         ...     ...               ...   
2827617  {F5E8B080-DE98-3A13-E053-6C04A8C060B7}  290000  2023-01-23 00:00   
2827618  {F3B6C198-1645-6E40-E053-6C04A8C0B3B4}  278000  2023-01-06 00:00   
2827619  {F3B6C198-1666-6E40-E053-6C04A8C0B3B4}  295000  2023-01-04 00:00   
2827620  {F3B6C198-166F-6E40-E053-6C04A8C0B3B4}  510000  2023-01-09 00:00   
2827621  {F3B6C198-1766-6E40-E053-6C04A8C0B3B4}  492500  2023-01-20 00:00   

         postcode Property_Type Old/New Duration         PAON      SAON  \
0         BR3 4AT             T       N        F           56       NaN   
1         UB5 5LH             S       N        F           57       NaN   
2        NN15 6UQ             S       N        F           80       NaN   
3         OX7 5RL             S       N        F            1       NaN   
4         NG2 6RP             D       Y        F            9       NaN   
...           ...           ...     ...      ...          ...       ...   
2827617    W6 0RP             F       N        L  KINGS COURT  FLAT 166   
2827618  CM19 5EE             F       N        L          65A       NaN   
2827619   CB9 7DS             D       N        F           51       NaN   
2827620   SS6 9EA             D       N        F           20       NaN   
2827621   CO6 2LG             S       N        F           16       NaN   

                 Street           Locality        Town/City  \
0             EDEN ROAD          BECKENHAM        BECKENHAM   
1         BENGARTH ROAD           NORTHOLT         NORTHOLT   
2            GOTCH ROAD    BARTON SEAGRAVE        KETTERING   
3             SUNNYBANK    GREAT ROLLRIGHT  CHIPPING NORTON   
4             THIRLMERE     WEST BRIDGFORD       NOTTINGHAM   
...                 ...                ...              ...   
2827617  HAMLET GARDENS                NaN           LONDON   
2827618     HIGH STREET             ROYDON           HARLOW   
2827619      WATER LANE  STEEPLE BUMPSTEAD        HAVERHILL   
2827620   LEONARD DRIVE                NaN         RAYLEIGH   
2827621        TEY ROAD        EARLS COLNE       COLCHESTER   

                       District            County PPDCategory_Type  \
0                       BROMLEY    GREATER LONDON                A   
1                        EALING    GREATER LONDON                A   
2                     KETTERING  NORTHAMPTONSHIRE                A   
3              WEST OXFORDSHIRE       OXFORDSHIRE                A   
4                    RUSHCLIFFE   NOTTINGHAMSHIRE                A   
...                         ...               ...              ...   
2827617  HAMMERSMITH AND FULHAM    GREATER LONDON                A   
2827618           EPPING FOREST             ESSEX                A   
2827619               BRAINTREE             ESSEX                A   
2827620                ROCHFORD             ESSEX                A   
2827621               BRAINTREE             ESSEX                A   

        Record_Status - monthly_file_only  
0                                       A  
1                                       A  
2                                       A  
3                                       A  
4                                       A  
...                                   ...  
2827617                                 A  
2827618                                 A  
2827619                                 A  
2827620                                 A  
2827621                                 A  

[2827622 rows x 16 columns]

In [ ]:
df.dropna(inplace=True)

In [ ]:
drop_duplicates(df)

No duplicate rows were found.


In [ ]:
df

Transaction_unique_identifier   price  Date_of_Transfer  \
10       {E76CCDAE-3A80-434F-99C3-91F5106F74E3}   35200  1995-06-09 00:00   
11       {5F8F69D0-70CA-45E8-8EA2-83B62CEABEA7}   88000  1995-11-15 00:00   
32       {F6EE4111-A464-45CB-88A9-5A08623CD6DB}  100500  1995-09-25 00:00   
36       {B79C45C6-8975-4735-B0AE-4F4B85CEB769}   95000  1995-10-17 00:00   
43       {F238A73D-8004-48B1-87F3-873969E4B3D7}   38000  1995-11-02 00:00   
...                                         ...     ...               ...   
2827440  {F5E8B081-161C-3A13-E053-6C04A8C060B7}  230000  2023-02-06 00:00   
2827441  {F5E8B081-1634-3A13-E053-6C04A8C060B7}  159000  2023-01-12 00:00   
2827442  {F5E8B081-1636-3A13-E053-6C04A8C060B7}  400000  2023-02-08 00:00   
2827530  {F5E8B081-16CE-3A13-E053-6C04A8C060B7}  630000  2023-02-02 00:00   
2827535  {F5E8B081-16FA-3A13-E053-6C04A8C060B7}  141500  2023-01-27 00:00   

         postcode Property_Type Old/New Duration                  PAON  \
10        SG7 5LW             F       N        L          THE MALTINGS   
11       TN14 7JP             T       N        F     ASH TREE COTTAGES   
32       TW11 9HX             F       N        L   THE OLD VICARAGE, 2   
36       WC2H 8DN             F       N        L                    30   
43       SO52 9AT             F       N        L      ST GEORGES COURT   
...           ...           ...     ...      ...                   ...   
2827440   BH2 5JA             F       N        L    BAY VIEW MEWS, 14A   
2827441  BH13 6HP             F       N        L          OAKHURST, 14   
2827442   BH6 3NP             F       N        L  THE SHELL HOUSE, 137   
2827530   DT2 8NE             S       N        F        UPTON COTTAGES   
2827535   BH5 1HD             F       N        L                    28   

            SAON                       Street         Locality    Town/City  \
10       FLAT 11                   GREEN LANE          ASHWELL      BALDOCK   
11             3              OLD LONDON ROAD        KNOCKHOLT    SEVENOAKS   
32        FLAT 2                KINGSTON ROAD       TEDDINGTON   TEDDINGTON   
36        FLAT 2           NEW COMPTON STREET           LONDON       LONDON   
43       FLAT 23                 BRACKEN ROAD  NORTH BADDESLEY  SOUTHAMPTON   
...          ...                          ...              ...          ...   
2827440  FLAT 12              WEST CLIFF ROAD       WESTBOURNE  BOURNEMOUTH   
2827441  FLAT 51                   THE AVENUE   BRANKSOME PARK        POOLE   
2827442   FLAT 8  SOUTHBOURNE OVERCLIFF DRIVE      SOUTHBOURNE  BOURNEMOUTH   
2827530        4                        UPTON        RINGSTEAD   DORCHESTER   
2827535   FLAT 7                  WESTBY ROAD         BOSCOMBE  BOURNEMOUTH   

                                    District  \
10                       NORTH HERTFORDSHIRE   
11                                 SEVENOAKS   
32                      RICHMOND UPON THAMES   
36                                    CAMDEN   
43                               TEST VALLEY   
...                                      ...   
2827440  BOURNEMOUTH, CHRISTCHURCH AND POOLE   
2827441  BOURNEMOUTH, CHRISTCHURCH AND POOLE   
2827442  BOURNEMOUTH, CHRISTCHURCH AND POOLE   
2827530                               DORSET   
2827535  BOURNEMOUTH, CHRISTCHURCH AND POOLE   

                                      County PPDCategory_Type  \
10                             HERTFORDSHIRE                A   
11                                      KENT                A   
32                            GREATER LONDON                A   
36                            GREATER LONDON                A   
43                                 HAMPSHIRE                A   
...                                      ...              ...   
2827440  BOURNEMOUTH, CHRISTCHURCH AND POOLE                A   
2827441  BOURNEMOUTH, CHRISTCHURCH AND POOLE                A   
2827442  BOURNEMOUTH, CHRISTCHURCH AND POOLE                A   
2827530                     

In [ ]:
cleaned_file_path = r"clean_202304.csv"
df.to_csv(cleaned_file_path, index=False)


# CHUNKING
- Iterates through chunks of the CSV file using pd.read_csv with the specified chunk_size. <br>
For each chunk: <br>
- Measures memory usage before and after processing the chunk using psutil.
- Records the start time and end time to calculate the computation time.
- Prints information about the current chunk, including shape, file size, memory usage, and computation time.
- Saves the processed chunk to a new CSV file.

In [ ]:
import pandas as pd
import os
import psutil
from time import time

cleaned_file_path = r"clean_202304.csv"

chunk_size = 10000
for i, chunk in enumerate(pd.read_csv(cleaned_file_path, chunksize=chunk_size)):

    memory_before = psutil.Process().memory_info().rss

    start_time = time()

    print(f"Chunk {i+1} Shape: {chunk.shape}")


    if i == 0:
        print(f"\nContents of Chunk {i+1}:\n{chunk}")


    chunk.to_csv(f"{cleaned_file_path}_chunk_{i+1}.csv", index=False)

    end_time = time()
    memory_after = psutil.Process().memory_info().rss
    computation_time = end_time - start_time

    file_size = os.path.getsize(f"{cleaned_file_path}_chunk_{i+1}.csv")
    print(f"Chunk {i+1} File Size: {file_size} bytes")
    print(f"Chunk {i+1} Memory Usage: {memory_after - memory_before} bytes")
    print(f"Chunk {i+1} Computation Time: {computation_time} seconds")

    if i == 0:
        print(f"\nContents of Chunk {i+1}:\n{chunk}")


Chunk 1 Shape: (10000, 16)

Contents of Chunk 1:
               Transaction_unique_identifier   price  Date_of_Transfer  \
0     {E76CCDAE-3A80-434F-99C3-91F5106F74E3}   35200  1995-06-09 00:00   
1     {5F8F69D0-70CA-45E8-8EA2-83B62CEABEA7}   88000  1995-11-15 00:00   
2     {F6EE4111-A464-45CB-88A9-5A08623CD6DB}  100500  1995-09-25 00:00   
3     {B79C45C6-8975-4735-B0AE-4F4B85CEB769}   95000  1995-10-17 00:00   
4     {F238A73D-8004-48B1-87F3-873969E4B3D7}   38000  1995-11-02 00:00   
...                                      ...     ...               ...   
9995  {0A8CD047-EC96-4B9A-AB5F-77C09E9D228B}   75000  1996-10-25 00:00   
9996  {9B66BE85-233B-4932-AB99-612ED359B78C}  135000  1996-07-30 00:00   
9997  {68C2ECD8-8077-4D0A-A1C1-7B55E4A37ADB}    8950  1996-01-26 00:00   
9998  {0FBBCBB9-10A8-451C-9F09-38551C51B7B9}   39950  1996-12-11 00:00   
9999  {DFA29D0D-B7B2-4A12-80CA-684A835D2277}   56500  1996-05-07 00:00   

      postcode Property_Type Old/New Duration                 

#SAMPLING
- Initializes variables to measure memory usage before and after the sampling process and record the start time for computation time.
- Iterates through each row of the CSV file using pd.read_csv with chunksize=1.
- For the first n rows, appends each row to the sampled_rows list.
- For subsequent rows (beyond the first n), calculates the probability of replacing an existing sampled row with the new row based on a uniform distribution.
- If the new row is selected, a randomly chosen existing row is replaced in the sampled_rows list.
- Calculates memory usage after the sampling process and the total computation time.

In [ ]:
n = 10000

memory_before = psutil.Process().memory_info().rss
start_time = time()

sampled_rows = []
for i, row in enumerate(pd.read_csv('clean_202304.csv', chunksize=1)):
    if i < n:
        sampled_rows.append(row)
    else:
        probability = n / (i + 1)
        if random.random() < probability:
            replace_index = random.randint(0, n - 1)
            sampled_rows[replace_index] = row

memory_after = psutil.Process().memory_info().rss
computation_time = time() - start_time

sampled_df = pd.concat(sampled_rows, axis=0, ignore_index=True)

print("Sampled DataFrame:")
print(sampled_df)

print(f"Memory Usage: {memory_after - memory_before} bytes")
print(f"Computation Time: {computation_time} seconds")


Sampled DataFrame:
               Transaction_unique_identifier   price  Date_of_Transfer  \
0     {367C2D15-FB4D-4B48-97BA-14DB489B63C3}  112000  2006-05-12 00:00   
1     {6A50285F-BD15-4C80-8DFF-890F2B4BC9EA}  136950  2004-02-27 00:00   
2     {3914047A-8805-3206-E050-A8C063057647}   60000  2016-06-29 00:00   
3     {FA73092C-4CCC-412B-8721-C7E6EF3297C4}  184000  2007-12-21 00:00   
4     {C8C2AC26-1135-4D59-A5C6-75C15754C9E9}   55000  1997-03-21 00:00   
...                                      ...     ...               ...   
9995  {B434581F-3564-4F5E-903D-DC250C22FD5A}  173000  2007-01-12 00:00   
9996  {1C9D4AFF-022C-4CEF-BDDE-ACF0F75BFBA9}  360000  2004-08-24 00:00   
9997  {68C2ECD8-8077-4D0A-A1C1-7B55E4A37ADB}    8950  1996-01-26 00:00   
9998  {5DF311FE-D3C3-4C11-AFC2-6B651804AF4A}   64000  1998-11-27 00:00   
9999  {13CE2011-2FAB-4F67-8389-B1B16A03A1CE}  193950  2007-02-23 00:00   

      postcode Property_Type Old/New Duration               PAON         SAON  \
0      LS1 

In [ ]:
sampled_df

Transaction_unique_identifier   price  Date_of_Transfer  \
0     {367C2D15-FB4D-4B48-97BA-14DB489B63C3}  112000  2006-05-12 00:00   
1     {6A50285F-BD15-4C80-8DFF-890F2B4BC9EA}  136950  2004-02-27 00:00   
2     {3914047A-8805-3206-E050-A8C063057647}   60000  2016-06-29 00:00   
3     {FA73092C-4CCC-412B-8721-C7E6EF3297C4}  184000  2007-12-21 00:00   
4     {C8C2AC26-1135-4D59-A5C6-75C15754C9E9}   55000  1997-03-21 00:00   
...                                      ...     ...               ...   
9995  {B434581F-3564-4F5E-903D-DC250C22FD5A}  173000  2007-01-12 00:00   
9996  {1C9D4AFF-022C-4CEF-BDDE-ACF0F75BFBA9}  360000  2004-08-24 00:00   
9997  {68C2ECD8-8077-4D0A-A1C1-7B55E4A37ADB}    8950  1996-01-26 00:00   
9998  {5DF311FE-D3C3-4C11-AFC2-6B651804AF4A}   64000  1998-11-27 00:00   
9999  {13CE2011-2FAB-4F67-8389-B1B16A03A1CE}  193950  2007-02-23 00:00   

      postcode Property_Type Old/New Duration               PAON         SAON  \
0      LS1 4ED             F       N        L                  3      FLAT 11   
1      M11 4AU             F       Y        L                  4  APARTMENT 4   
2      DN4 6BX             F       N        L                 51       FLAT 2   
3     KT16 8DE             F       N        L  BOURNE PLACE, 101      FLAT 23   
4       E5 9AJ             F       N        L        TOWER COURT       FLAT 2   
...        ...           ...     ...      ...                ...          ...   
9995   SL3 8PP             F       N        L       GIBSON COURT           14   
9996   BR3 1DQ             F       Y        L      GALLERY HOUSE       FLAT 9   
9997   BA6 9LT             F       N        L   OLD MARKET COURT      FLAT 45   
9998  CV31 1BQ             F       N        L                 37       FLAT 2   
9999  BN27 1BH             F       Y        L      GARRETT COURT      FLAT 26   

                     Street        Locality       Town/City    District  \
0     LITTLE NEVILLE STREET           LEEDS           LEEDS       LEEDS   
1            THE WATERFRONT      MANCHESTER      MANCHESTER  MANCHESTER   
2        GOODISON BOULEVARD         CANTLEY       DONCASTER   DONCASTER   
3            EASTWORTH ROAD        CHERTSEY        CHERTSEY   RUNNYMEDE   
4            CLAPTON COMMON          LONDON          LONDON     HACKNEY   
...                     ...             ...             ...         ...   
9995            DITTON ROAD          SLOUGH          SLOUGH      SLOUGH   
9996       COPERS COPE ROAD       BECKENHAM       BECKENHAM     BROMLEY   
9997          GEORGE STREET     GLASTONBURY     GLASTONBURY      MENDIP   
9998           LEAM TERRACE  LEAMINGTON SPA  LEAMINGTON SPA     WARWICK   
9999          VICARAGE LANE        HAILSHAM        HAILSHAM     WEALDEN   

                  County PPDCategory_Type Record_Status - monthly_file_only  
0         WEST YORKSHIRE                A                                 A  
1     GREATER MANCHESTER                A                                 A  
2        SOUTH YORKSHIRE                B                                 A  
3                 SURREY                A                                 A  
4         GREATER LONDON                A                                 A  
...                  ...              ...                               ...  
9995              SLOUGH                A                                 A  
9996      GREATER LONDON                A                                 A  
9997            SOMERSET                A                                 A  
9998        WARWICKSHIRE                A                                 A  
9999         EAST SUSSEX                A                                 A  

[10000 rows x 16 columns]

# LOAD SUBSET OF DATA FROM THE DATASET
- Uses the pd.read_csv function to read the CSV file, specifying the selected columns using the usecols parameter.
- Creates a DataFrame (less_data_df) containing only the selected columns.
- Prints the first few rows of the DataFrame using head() to provide a preview of the data.
- Prints summary information about the DataFrame using info(), including the data types and non-null counts for each column.

In [ ]:
selected_columns = ["price", "Property_Type", "Old/New", "Town/City", "County"]

file_path = 'clean_202304.csv'
less_data_df = pd.read_csv(file_path, usecols = selected_columns)

print(less_data_df.head())
less_data_df.info()

    price Property_Type Old/New    Town/City          County
0   35200             F       N      BALDOCK   HERTFORDSHIRE
1   88000             T       N    SEVENOAKS            KENT
2  100500             F       N   TEDDINGTON  GREATER LONDON
3   95000             F       N       LONDON  GREATER LONDON
4   38000             F       N  SOUTHAMPTON       HAMPSHIRE
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155139 entries, 0 to 155138
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   price          155139 non-null  int64 
 1   Property_Type  155139 non-null  object
 2   Old/New        155139 non-null  object
 3   Town/City      155139 non-null  object
 4   County         155139 non-null  object
dtypes: int64(1), object(4)
memory usage: 5.9+ MB


In [ ]:
print("Memory usage for load less data:", less_data_df.memory_usage().sum() / (1024 ** 2), "MB")

import timeit
time = timeit.timeit(lambda:less_data_df, number=1)
print("Computation time for load less data:", time, "seconds")

Memory usage for load less data: 5.918205261230469 MB
Computation time for load less data: 1.2990000186619e-06 seconds


# PARALLELIZE WITH DASK
- Uses dask.dataframe.read_csv to create a Dask DataFrame (dask_df) from the CSV file. Dask allows parallelized computation on larger-than-memory datasets.
- Measures memory usage before the computation using psutil.
- Records the start time for computation time measurement.
- Performs a groupby operation on the 'Property_Type' column, calculating the mean of the 'price' column using Dask's parallelized computing capabilities.
- Uses the compute() method to trigger the computation and retrieve the result as a pandas Series (dask_result).
- Records the end time for computation time measurement.
- Measures memory usage after the computation using psutil.

In [ ]:
import dask.dataframe as dd
import psutil
from time import time

dask_df = dd.read_csv('clean_202304.csv')

memory_before = psutil.Process().memory_info().rss

start_time = time()

dask_result = dask_df.groupby('Property_Type')['price'].mean().compute()

end_time = time()
memory_after = psutil.Process().memory_info().rss
computation_time = end_time - start_time

print("Dask Result:")
print(dask_result)

print(f"Memory Usage: {memory_after - memory_before} bytes")
print(f"Computation Time: {computation_time} seconds")


Dask Result:
Property_Type
D    340935.633725
F    169931.895308
O    731310.689959
S    220591.355450
T    191138.377519
Name: price, dtype: float64
Memory Usage: 68833280 bytes
Computation Time: 0.6401019096374512 seconds


# OPTIMIZE DATA TYPES

- Uses dask.dataframe.read_csv to create a Dask DataFrame (dask_df) from the CSV file.
- Converts specific columns to more memory-efficient data types to optimize memory usage:
'price' to 'float32'.
'Date_of_Transfer' and categorical columns to 'category'.
- Measures memory usage before the fine-tuning process using psutil.
- Records the start time for computation time measurement (though there is no significant computation in this section).
- Records the end time for computation time measurement.
- Measures memory usage after the fine-tuning process using psutil.
- Calculates the total computation time for the fine-tuning process.

In [ ]:
import dask.dataframe as dd
import psutil
from time import time


dask_df = dd.read_csv('clean_202304.csv')

dask_df['price'] = dask_df['price'].astype('float32')
dask_df['Date_of_Transfer'] = dask_df['Date_of_Transfer'].astype('category')
dask_df['Property_Type'] = dask_df['Property_Type'].astype('category')
dask_df['Old/New'] = dask_df['Old/New'].astype('category')
dask_df['postcode'] = dask_df['postcode'].astype('category')
dask_df['Duration'] = dask_df['Duration'].astype('category')
dask_df['PAON'] = dask_df['PAON'].astype('category')
dask_df['SAON'] = dask_df['SAON'].astype('category')
dask_df['Locality'] = dask_df['Locality'].astype('category')
dask_df['Town/City'] = dask_df['Town/City'].astype('category')
dask_df['District'] = dask_df['District'].astype('category')
dask_df['County'] = dask_df['County'].astype('category')
dask_df['PPDCategory_Type'] = dask_df['PPDCategory_Type'].astype('category')

memory_before_processing = psutil.Process().memory_info().rss

start_time_processing = time()

end_time_processing = time()
memory_after_processing = psutil.Process().memory_info().rss
computation_time_processing = end_time_processing - start_time_processing

print(f"Memory Usage (After Fine-Tuning): {memory_after_processing - memory_before_processing} bytes")
print(f"Computation Time (After Fine-Tuning): {computation_time_processing} seconds")

fine_tuned_data_types = dask_df.dtypes
print("\nFine-Tuned Data Types:")
print(fine_tuned_data_types)


Memory Usage (After Fine-Tuning): 0 bytes
Computation Time (After Fine-Tuning): 2.9802322387695312e-05 seconds

Fine-Tuned Data Types:
Transaction_unique_identifier          object
price                                 float32
Date_of_Transfer                     category
postcode                             category
Property_Type                        category
Old/New                              category
Duration                             category
PAON                                 category
SAON                                 category
Street                                 object
Locality                             category
Town/City                            category
District                             category
County                               category
PPDCategory_Type                     category
Record_Status - monthly_file_only      object
dtype: object


# COMPARATIVE ANALYSIS

### Result and Analysis

| Strategy                | Memory Used (MB) | Computation Time      | Analysis                        |
|-------------------------|------------------|-----------------------|---------------------------------------------------------------------|
| Load Less Data          | 5.92             | 1.2990 microseconds   |   Being selective about the data which is loaded for analysis (e.g. only three columns) resulted in the quickest computation time, very useful for specific usecases especially for known output retrieval                       |
| Chunking                | 0.00             | 1.5094 seconds (sum)  | Chunking is negligible in terms of CPU memory and very fast in terms of computation time, it is most useful in working with big datasets, even with memory limitations                                                   |
| Data Type Optimization  | 0.00             | 29.8023 microseconds  |      It is very a simple method that results in noticeable computation time improvements at basically no other cost                              |
| Sampling                | 33.60            | 312.3491 seconds      | Effects on memory usage and computation time depends on implementation sampling i.e. randomly sampling dataset or focusing on specific parts yields different costs, our example is very time inducing for example, which might not be as worth it                                                   |
| Parallelize with Dask   | 68.83            | 0.6401 seconds        |    Even though this strategy utilises the most memory, it is remarkably effective when implemented on very large datasets, reducing computational workloads, enhancing responsiveness and retaining insight of entire dataset|

### Comparison Between Five Strategies

#### In Terms of Approach:
| Strategy                | Traditional Approach | Advanced Approach      
|-------------------------|------------------|-----------------------|
| Load Less Data          | Loads entire dataset into memory thus very memory intensive and less optimised | Only most relevant data are selected to load into memory |
| Chunking                | Processes all data in one go | Breaks data into smaller, manageable parts | Memory-efficient processing in smaller pieces |
| Data Type Optimization  | Automatically infers data types | Fine-tunes data types for efficiency (e.g. data type size)|
| Sampling                | Analyzes the entire dataset | Analyzes a representative subset (e.g., randomly or percentile of dataset) |
| Parallelize processing using the Dask library  | Lacks support for parallel computing | Uses Dask for parallel and distributed computing	|

#### In Terms of Pros and Cons:
| Strategy                | Pros | Cons
|-------------------------|----------|-------|
| Load Less Data          | Minimising memory usage, greatly improving handling of large data, making it possible to analyze big datasets | Limits the amount of data available for information gathering |
| Chunking                | Memory-efficient processing in smaller pieces | Some operations may require additional handling across chunks |
| Data Type Optimization  | Reduces memory usage significantly by using more compact data types | Careful selection of columns is crucial based on data distribution |
| Sampling                | Speeds up analysis, especially when a full-scale examination is not needed, can also be used to quickly analyse random or representative subset of dataset | May miss specific patterns or cannot capture rare occurrences present in the entire dataset |
| Parallelize with Dask   | Drastically cuts down computation time, enables working with datasets larger than memory | Requires Dask installation and adjustment of computations for parallelism |


# CONCLUSION


In conclusion, the choice of strategy in working with big data significantly influences the efficiency, resource utilization, and overall success of data analysis. Each strategy comes with its own set of benefits and drawbacks.

1. **Load Less Data**  
    
    Stands out for its emphasis on minimizing memory usage, thereby enhancing the handling of large datasets. However, it comes at the cost of limiting the amount of data available for analysis, potentially sacrificing comprehensive insights.

2. **Chunking**
    
    Addresses memory challenges by breaking down the data into manageable parts. While this offers memory-efficient processing, it introduces complexities in handling operations across chunks, which may require additional considerations.

3. **Data Type Optimization**
    
    Advantageous for reducing memory usage by employing more compact data types. However, careful column selection based on data distribution is crucial for its success.

4. **Sampling**
    
    Accelerates analysis by examining a representative subset of the dataset, but it runs the risk of missing specific patterns or failing to capture rare occurrences present in the entire dataset.

5. **Parallelize with Dask**
    
    Drastically cuts down computation time and enabling the analysis of datasets larger than memory. However, it demands the installation of Dask and adjustments to computations for parallelism.

In summary, the selection of a strategy should align with the specific requirements of the analysis, considering factors such as data size, memory constraints, and the need for comprehensive insights. An informed choice based on these considerations will contribute to more efficient and effective big data analysis.